Add build-and-analyze and generate-static-analysis-archive for static analysis

briannafan opened the pull request:
#### 723136dec2e8a374697f168760208919b88a4bd6
<pre>
Add build-and-analyze and generate-static-analysis-archive for static analysis
<a href="https://bugs.webkit.org/show_bug.cgi?id=268862">https://bugs.webkit.org/show_bug.cgi?id=268862</a>
<a href="https://rdar.apple.com/problem/122427584">rdar://problem/122427584</a>

Reviewed by David Kilzer.

Added scripts with minor changes.

* Tools/Scripts/build-and-analyze: Added.
(args_for_additional_checkers):
(make_analyzer_flags):
(scan_build_path):
(main):
(parse_args): Added &apos;--scan-build-dir&apos; flag to override default scan-build path.
* Tools/Scripts/generate-static-analysis-archive: Added.
(parse_command_line):
(get_project_name):
(generate_results_page):
(get_total_issue_count):
(main):
* Tools/Scripts/webkitpy/static_analysis/__init__.py: Added.
* Tools/Scripts/webkitpy/static_analysis/results.py: Added.
(get_project_issue_count_as_string): Used in generate-static-analysis-archive.
* Tools/Scripts/webkitpy/static_analysis/results_unittest.py: Added.
(StaticAnalysisResultsTest):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_no_file):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_invalid_long):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_invalid_short):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_valid):

Canonical link: <a href="https://commits.webkit.org/274239@main">https://commits.webkit.org/274239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e287918ae5c40e27690ce3bed90d78672d32c500

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38361 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34111 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40656 "Failed to checkout and rebase branch from PR 23961") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14642 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32335 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14609 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12707 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/38254 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42197 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38527 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36727 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14838 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4998 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14302 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/Scripts/build-and-analyze:
```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+# Copyright (C) 2024 Apple Inc. All rights reserved.
```
Nit:  Technically, this should be `2023-2024`.  Not sure if it matters in practice.

ddkilzer commented on Tools/Scripts/build-and-analyze:
```diff
@@ -0,0 +1,195 @@
+def scan_build_path(sdkroot):
+    """
+    Returns path to scan-build from the macosx.internal SDK.
+    The internal SDK must be used due to <rdar://53628544>.
```
Change this comment to something like:

```suggestion
    Try to find scan-build in the SDK, else assume scan-build is in the
    user's PATH.  It may be downloaded from the clang/llvm repository.
```

ddkilzer commented on Tools/Scripts/build-and-analyze:
```diff
@@ -0,0 +1,195 @@
+                ['xcrun', '-find', '-sdk', sdkroot, 'scan-build'],
+                stderr=dev_null).rstrip().decode('utf-8'))
+    except subprocess.CalledProcessError:
+        return None
```
Change to:

```suggestion
        # FIXME: Replace scan-build with our own script, or add a copy to WebKit.
        return 'scan-build'
```

ddkilzer commented on Tools/Scripts/build-and-analyze:
```diff
@@ -0,0 +1,195 @@
+    parser.add_argument('--output-dir', dest='output_dir', required=True,
+                        help='Output directory for HTML results.')
+    parser.add_argument('--sdkroot', dest='sdkroot', default='macosx.internal',
+                        help='SDKROOT to use (default: macosx.internal).')
```
Change the default from `macosx.internal` to `macosx`.

ddkilzer commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python
```
Can we update this script to use `python3` as well?  If not, we can do that later.

ddkilzer commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python
+# Copyright (C) 2024 Apple Inc. All rights reserved.
```
Nit: This should be `2014-2024`.

ddkilzer commented on Tools/Scripts/webkitpy/static_analysis/results.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright (C) 2024 Apple Inc. All rights reserved.
```
Nit:  This should be `2020-2024`.

ddkilzer reviewed (approved):
Looks good!  I think we can land this after making the requested changes.

briannafan committed:
Add build-and-analyze and generate-static-analysis-archive for static analysis
https://bugs.webkit.org/show_bug.cgi?id=268862
rdar://problem/122427584

Reviewed by David Kilzer.

Added scripts with minor changes.

* Tools/Scripts/build-and-analyze: Added.
(args_for_additional_checkers):
(make_analyzer_flags):
(scan_build_path):
(main):
(parse_args): Added '--scan-build-dir' flag to override default scan-build path.
* Tools/Scripts/generate-static-analysis-archive: Added.
(parse_command_line):
(get_project_name):
(generate_results_page):
(get_total_issue_count):
(main):
* Tools/Scripts/webkitpy/static_analysis/__init__.py: Added.
* Tools/Scripts/webkitpy/static_analysis/results.py: Added.
(get_project_issue_count_as_string): Used in generate-static-analysis-archive.
* Tools/Scripts/webkitpy/static_analysis/results_unittest.py: Added.
(StaticAnalysisResultsTest):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_no_file):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_invalid_long):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_invalid_short):
(StaticAnalysisResultsTest.test_get_project_issue_count_as_string_valid):

Canonical link: https://commits.webkit.org/274239@main
